kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/MergeQueries.java

```diff
@@ -0,0 +1,336 @@
+  public static final String INTERMEDIATE_TABLE_BATCH_NUMBER_FIELD = "batchNumber";
+
+  private static final Logger logger = LoggerFactory.getLogger(MergeQueries.class);
+
+  private final String keyFieldName;
+  private final boolean usePartitionDecorator;
+  private final boolean upsertEnabled;
+  private final boolean deleteEnabled;
+  private final MergeBatches mergeBatches;
+  private final ExecutorService executor;
+  private final BigQuery bigQuery;
+  private final SchemaManager schemaManager;
+  private final SinkTaskContext context;
+
+  public MergeQueries(BigQuerySinkTaskConfig config,
+                      MergeBatches mergeBatches,
+                      ExecutorService executor,
+                      BigQuery bigQuery,
+                      SchemaManager schemaManager,
+                      SinkTaskContext context) {
+    this.keyFieldName = config.getKafkaKeyFieldName().orElseThrow(() ->
+        new ConnectException("Kafka key field must be configured when upsert/delete is enabled")
+    );
+    this.usePartitionDecorator = config.getBoolean(config.BIGQUERY_PARTITION_DECORATOR_CONFIG);
+    this.upsertEnabled = config.getBoolean(config.UPSERT_ENABLED_CONFIG);
+    this.deleteEnabled = config.getBoolean(config.DELETE_ENABLED_CONFIG);
+
+    this.mergeBatches = mergeBatches;
+    this.executor = executor;
+    this.bigQuery = bigQuery;
+    this.schemaManager = schemaManager;
+    this.context = context;
+  }
+
+  public void mergeFlushAll() {
+    logger.debug("Triggering merge flush for all tables");
+    mergeBatches.intermediateTables().forEach(this::mergeFlush);
+  }
+
+  public void mergeFlush(TableId intermediateTable) {
+    final TableId destinationTable = mergeBatches.destinationTableFor(intermediateTable);
+    final int batchNumber = mergeBatches.incrementBatch(intermediateTable);
+    logger.trace("Triggering merge flush from intermediate table {} to destination table {} for batch {}",
+        intermediateTable, destinationTable, batchNumber);
+
+    executor.submit(() -> {
+      // If there are rows to flush in this batch, flush them
+      if (mergeBatches.prepareToFlush(intermediateTable, batchNumber)) {
+        try {
+          logger.debug("Running merge query on batch {} from intermediate table {}",
+              batchNumber, intermediateTable);
+          String mergeFlushQuery = mergeFlushQuery(intermediateTable, destinationTable, batchNumber);
+          logger.trace(mergeFlushQuery);
+          bigQuery.query(QueryJobConfiguration.of(mergeFlushQuery));
+          logger.trace("Merge from intermediate table {} to destination table {} completed",
+              intermediateTable, destinationTable);
+        } catch (Throwable t) {
+          logger.warn("Failed on merge flush from intermediate table {} to destination table {}",
+              intermediateTable, destinationTable, t);
+          throw new ConnectException(
```

Comment:
This exception actually will get swallowed unless you handle the `ExecutionException` thrown on `Future.get`. https://stackoverflow.com/a/39091808/12414477

Looks like the executor implementation used overrides `afterExecute` hook. You can update it to catch both types of exceptions like this - https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/ThreadPoolExecutor.html#afterExecute(java.lang.Runnable,%20java.lang.Throwable)

Reply:
Yep, this is exactly what the connector does already :)

https://github.com/wepay/kafka-connect-bigquery/blob/c4d76ea5a1519aae4a42c22e2d3989caee8a787b/kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/write/batch/KCBQThreadPoolExecutor.java#L63-L74

I'll pull out the `Runnable` into a separate method to be clearer what's being invoked by the `executor` field, and make the type signature of that field more specific so that it's obvious that it's an instance of `KCBQThreadPoolExecutor`.

Reply:
The part that I'm worried about is this blurb from javadoc:
```
Note: When actions are enclosed in tasks (such as FutureTask) either explicitly or via methods such as submit, these task objects catch and maintain computational exceptions, and so they do not cause abrupt termination, and the internal exceptions are not passed to this method.
```
This does not seem to be handled in `KCBQThreadPoolExecutor:: afterExecute `

Reply:
🤦‍♀️ no kidding. I should use `execute` here instead of `submit`. Good catch, thanks Nigel!